### PR TITLE
[TechDocs] Experimental useTechDocsReaderDom hook

### PR DESCRIPTION
**Pull request by camilaibs**
## What / Why

Although it is possible to extend TechDocs at build-time (using, for example, the Mkdocs plugin ecosystem), it is not currently possible to extend TechDocs with _dynamic_ information available only at _read-time_.

We're not 100% sure what an API for providing such functionality should look like, but we'd like to start experimenting.

Our proposal in this PR is to refactor the TechDocs `<Reader />` component so that the core process of loading HTML from storage, sanitizing it, and making it work within the Backstage chrome, is encapsulated within a hook.

The reader would continue to behave exactly as it does now, leveraging this new hook.  However we'll also _export_ the hook so that we can try out alternative implementations which can extend TechDocs' functionality.  Such an implementation might look like:

```jsx
const CustomTechDocsReader = withTechDocsReaderProvider(() => {
  const dom = useTechDocsReaderDom();
  
  useEffect(() => {
    if (!dom || !ref.current) return;
    const shadowRoot = ref.current.attachShadow({ mode: 'open' });
    shadowRoot.appendChild(dom);
  }, [dom]);
  
  return <div ref={ref} />;
});
```

We also extracted and exported two other components to use in experimental Reader implementations:

 - TechDocsSearch
 - TechDocsStateIndicator

The hook and the above components are marked as `@internal` to clearly indicate that they should not be relied upon by other plugins.  We may choose to not export them in the future.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by emmaindal — COMMENTED** [review]
Reading the description of the PR one thing came to mind - Is this something we expect to stay? if it's currently an experiment I wonder if it makes sense to have an experimental Reader which uses the newly introduced hook and then leave the one used already untouched? what do you think?

File: plugins/techdocs/src/reader/components/Reader.tsx [hunk 1]
```diff
@@ -358,94 +310,48 @@ export const Reader = ({ entityRef, onReady, withSearch = true }: Props) => {
...
-              Building a newer version of this documentation failed.{' '}
-              {syncErrorMessage}
-            </Alert>
-          )}
-          <TechDocsNotFound errorMessage={contentErrorMessage} />
-        </>
-      )}
-
+      <TechDocsProgress />
+      <TechDocsAlert />
```
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/techdocs/src/reader/components/Reader.tsx:354]: The progress is a generic Loading component, and the Alert is a section on the screen where we show the status of the documentation build.
I can put them together.

File: plugins/techdocs/src/reader/components/Reader.tsx [hunk 2]
```diff
@@ -358,94 +310,48 @@ export const Reader = ({ entityRef, onReady, withSearch = true }: Props) => {
...
-        shadowRoot.removeChild(child),
-      );
-      shadowRoot.appendChild(transformedElement);
-
       // Scroll to top after render
       window.scroll({ top: 0 });
 
       // Post-render
-      await postRender(shadowRoot);
+      const renderedElement = await postRender(transformedElement);
```
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/techdocs/src/reader/components/Reader.tsx:317]: This is the DOM element after the pre-transformations

File: plugins/techdocs/src/reader/components/TechDocsAlert.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      </Alert>
+    );
+  }
+
+  if (state === 'CONTENT_STALE_REFRESHING') {
+    return (
+      <Alert
+        variant="outlined"
+        severity="info"
+        icon={<CircularProgress size="24px" />}
```
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/techdocs/src/reader/components/TechDocsAlert.tsx:66]: I think the idea behind the issue is to remove circular progress from the alert and just keep the progress bar.
It appears that the progress bar is only rendered when the status is `CHECKING`. I think we can show the progress bar to the same status used by the alert component and then remove the icon from there.
I will test it.

File: plugins/techdocs/api-report.md [hunk 4]
```diff
@@ -436,6 +461,11 @@ export class TechDocsStorageClient implements TechDocsStorageApi {
   ): Promise<SyncResult>;
 }
 
+// Warning: (ae-missing-release-tag) "useTechDocsReaderDom" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export const useTechDocsReaderDom: () => Element | null;
```
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/techdocs/api-report.md]: Done!

File: plugins/techdocs/src/reader/components/Reader.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/techdocs/src/reader/components/Reader.tsx:318]: Do you think the `preTransformedDomElement` and `postTransformedDomElement` are better names 🤔 ?

**Comment by iamEAP** [comment]
> Is this something we expect to stay? if it's currently an experiment I wonder if it makes sense to have an experimental Reader which uses the newly introduced hook and then leave the one used already untouched? what do you think?

Good question!  In any event, given how large the `<Reader />` component has grown since its inception, refactoring out some of the logic into smaller constituents _feels_ good?  It's a little easier to read through and understand.

So maybe one way to look at this is that the fact that the hook is _exported_ is experimental (we may choose to not export it in the future), and perhaps even some of the behavior within the hook could change too.  But we'll continue to ensure that the `<Reader />` component itself is stable and behaves the exact same as it does now.

**Pull request merged**